### PR TITLE
Replace standard `log` with logrus

### PR DESCRIPTION
**Pull request by pudelkoM**
This PR replaces the standard lib `log` module with the more flexible `logrus` one.
It allows for leveled logging, structured logging, colors and flexible output formats (json, text, ...).

The log level can be set in the json config file with field `log_level` and can take one of the following string values: `"panic", "fatal", "error", "warning", "info", "debug", "trace"`. See [`ParseLevel`](https://github.com/sirupsen/logrus/blob/fdf1618bf7436ec3ee65753a6e2999c335e97221/logrus.go#L25-L45) for details.

With more than one general log level, certain log messages too expensive to print in production have been enabled at `trace` severity again.

Fixes: #251

New log output:
```
INFO[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/main.go:145 main.main() {sim 50000 {ens803f2} {ens803f3} {   false 10.250.0.0/16 8080 } {172.17.0.1/32 onos 51001 10.250.0.0/24} false {16.0.0.1 11.1.1.129 13.1.1.199 6.6.6.6 9.9.9.9 0x30000000 0x90000000} 0 0 false false   info}
INFO[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/upf.go:105 main.(*upf).setUpfInfo() UPF Node IP :  0.0.0.0
INFO[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/upf.go:106 main.(*upf).setUpfInfo() UPF Local IP :  0.0.0.0
INFO[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/bess.go:302 main.(*bess).setUpfInfo() setUpfInfo bess
INFO[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/bess.go:306 main.(*bess).setUpfInfo() IP pool :  10.250.0.0/16
FATA[2021-08-05T14:24:26-07:00]/Users/max/projects/upf-epc/pfcpiface/main.go:109 main.ParseIP() Unable to get info on interface name: ens803f2 route ip+net: no such network interface
```

Old:
```
2021/08/05 14:12:57 main.go:133: { 50000 {ens803f2} {ens803f3} {   false 10.250.0.0/16 8080 } {172.17.0.1/32 onos 51001 10.250.0.0/24} false {16.0.0.1 11.1.1.129 13.1.1.199 6.6.6.6 9.9.9.9 0x30000000 0x90000000} 0 0 false false  }
2021/08/05 14:13:02 upf.go:105: UPF Node IP :  0.0.0.0
2021/08/05 14:13:02 upf.go:106: UPF Local IP :  0.0.0.0
2021/08/05 14:13:02 bess.go:302: setUpfInfo bess
2021/08/05 14:13:02 bess.go:306: IP pool :  10.250.0.0/16
2021/08/05 14:13:02 main.go:108: Unable to get info on interface name: ens803f2 route ip+net: no such network interface
```

### CODE REVIEW

**Comment by krsna1729** [comment]
There might be a better way to pretty print struct. If not lets put level trace here

https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-pdr.go#L50-L77
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-far.go#L55-L69
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-qer.go#L25-L38

And then uncomment lines calling it
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/bess.go#L75-L100

Mention this in your first comment/commit Fixes: #251

**Comment by krsna1729** [comment]
There might be a better way to pretty print struct. If not lets put level trace here

https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-pdr.go#L50-L77
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-far.go#L55-L69
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/parse-qer.go#L25-L38

And then uncomment lines calling it
https://github.com/omec-project/upf-epc/blob/master/pfcpiface/bess.go#L75-L100

Mention this in your first comment/commit Fixes: #251

**Comment by krsna1729** [comment]
I think we can replace all the print methods with just this?

```
log.Traceln("%+v\n", pdr)
log.Traceln("%+v\n", far)
log.Traceln("%+v\n", qer)

```

**Comment by krsna1729** [comment]
Cant because there are IP conversions

**Comment by pudelkoM** [comment]
> Cant because there are IP conversions

Same conclusion on my end, check latest commit for alternative.

File: pfcpiface/parse-far.go [hunk 1]
```diff
@@ -52,20 +54,21 @@ type far struct {
...
+	fmt.Println(&b, "fseID:", f.fseID)
+	fmt.Println(&b, "fseIDIP:", f.fseidIP)
+	fmt.Println(&b, "dstIntf:", f.dstIntf)
+	fmt.Println(&b, "applyAction:", f.applyAction)
+	fmt.Println(&b, "tunnelType:", f.tunnelType)
+	fmt.Println(&b, "tunnelIP4Src:", f.tunnelIP4Src)
+	fmt.Println(&b, "tunnelIP4Dst:", f.tunnelIP4Dst)
+	fmt.Println(&b, "tunnelTEID:", f.tunnelTEID)
+	fmt.Println(&b, "tunnelPort:", f.tunnelPort)
+	fmt.Println(&b, "sendEndMarker:", f.sendEndMarker)
```
Comment (pudelkoM) [inline pfcpiface/parse-far.go:70]: Have to check how this actually looks in a log message. I image logrus might strip `"\n"` characters.

Comment (krsna1729) [reply pfcpiface/parse-far.go:70]:
```
time="2021-08-06T04:56:10Z" level=info msg="create sessions: 1" func=main.main file="/pfcpiface/main.go:184"
 Src Iface: 2
 tunnelIP4Dst: 0.0.0.0
 tunnelTEID: 0
 srcIP: 0.0.0.0
 dstIP: 16.0.0.1
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 0.0.0.0
 tunnelTEIDMask Mask: 0
 srcIP Mask: 0.0.0.0
 dstIP Mask: 255.255.255.255
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 0
 qerID: 0
 needDecap: 0
 allocIPFlag: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:76"
 Src Iface: 2
 tunnelIP4Dst: 198.19.0.1
 tunnelTEID: 2415919104
 srcIP: 0.0.0.0
 dstIP: 0.0.0.0
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 0.0.0.0
 dstIP Mask: 0.0.0.0
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 0
 qerID: 0
 needDecap: 1
 allocIPFlag: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:76"
 Src Iface: 1
 tunnelIP4Dst: 198.18.0.1
 tunnelTEID: 805306368
 srcIP: 16.0.0.1
 dstIP: 0.0.0.0
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 255.255.255.255
 dstIP Mask: 0.0.0.0
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 1
 qerID: 1
 needDecap: 1
 allocIPFlag: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:76"
 Src Iface: 1
 tunnelIP4Dst: 198.18.0.1
 tunnelTEID: 805306368
 srcIP: 0.0.0.0
 dstIP: 9.9.9.9
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 0.0.0.0
 dstIP Mask: 255.255.255.255
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 2
 qerID: 2
 needDecap: 1
 allocIPFlag: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:76"
 FAR ID: 0
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 0
 applyAction: 2
 tunnelType: 1
 tunnelIP4Src: 3323068417
 tunnelIP4Dst: 184615297
 tunnelTEID: 805306368
 tunnelPort: 2152
 sendEndMarker: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:87"
 FAR ID: 1
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 1
 applyAction: 2
 tunnelType: 0
 tunnelIP4Src: 0
 tunnelIP4Dst: 0
 tunnelTEID: 0
 tunnelPort: 0
 sendEndMarker: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:87"
 FAR ID: 2
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 1
 applyAction: 2
 tunnelType: 1
 tunnelIP4Src: 3323133953
 tunnelIP4Dst: 218169799
 tunnelTEID: 2415919104
 tunnelPort: 2152
 sendEndMarker: false
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:87"
 QER ID: 0
 fseID: 805306368
 QFI: 9
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:98"
 QER ID: 1
 fseID: 805306368
 QFI: 8
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:98"
 QER ID: 2
 fseID: 805306368
 QFI: 7
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
time="2021-08-06T04:56:10Z" level=trace func="main.(*bess).sendMsgToUPF" file="/pfcpiface/bess.go:98"
time="2021-08-06T04:56:10Z" level=info msg="Sessions/s: 215.61543004389713" func="main.(*bess).sim" file="/pfcpiface/bess.go:537"
```

**Comment by krsna1729** [comment]
Can you run the below and update the PR

```
go mod tidy
go mod vendor
```

**Comment by pudelkoM** [comment]
> Can you run the below and update the PR
> 
> ```
> go mod tidy
> go mod vendor
> ```

Done.

File: pfcpiface/parse-far.go (hunk 1 above)
Comment (krsna1729) [reply pfcpiface/parse-far.go:70]: Lets make the text accompanying the struct member case consistent. Lets fix the formatting decimal/hex/IP as appropriate and consistent across all 3 - PDR/FAR/QER. Some that I could see

fseID
tunnelTEID
tunnelTEIDMask
tunnelIP4Src
tunnelIP4Dst
tunnelTEID

**Comment by krsna1729** [comment]
The above output was without `-it`. This is how it looks with `-it` option in docker

```
docker exec -it bess-pfcpiface pfcpiface -config /conf/upf.json -simulate create
```
```
INFO[2021-08-06T05:13:21Z]/pfcpiface/main.go:184 main.main() create sessions: 1                           
 Src Iface: 2
 tunnelIP4Dst: 0.0.0.0
 tunnelTEID: 0
 srcIP: 0.0.0.0
 dstIP: 16.0.0.1
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 0.0.0.0
 tunnelTEIDMask Mask: 0
 srcIP Mask: 0.0.0.0
 dstIP Mask: 255.255.255.255
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 0
 qerID: 0
 needDecap: 0
 allocIPFlag: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:76 main.(*bess).sendMsgToUPF()                                              
 Src Iface: 2
 tunnelIP4Dst: 198.19.0.1
 tunnelTEID: 2415919104
 srcIP: 0.0.0.0
 dstIP: 0.0.0.0
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 0.0.0.0
 dstIP Mask: 0.0.0.0
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 0
 qerID: 0
 needDecap: 1
 allocIPFlag: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:76 main.(*bess).sendMsgToUPF()                                              
 Src Iface: 1
 tunnelIP4Dst: 198.18.0.1
 tunnelTEID: 805306368
 srcIP: 16.0.0.1
 dstIP: 0.0.0.0
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 255.255.255.255
 dstIP Mask: 0.0.0.0
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 1
 qerID: 1
 needDecap: 1
 allocIPFlag: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:76 main.(*bess).sendMsgToUPF()                                              
 Src Iface: 1
 tunnelIP4Dst: 198.18.0.1
 tunnelTEID: 805306368
 srcIP: 0.0.0.0
 dstIP: 9.9.9.9
 srcPort: 0
 dstPort: 0
 proto: 0
 Src Iface Mask: 255
 tunnelIP4Dst Mask: 255.255.255.255
 tunnelTEIDMask Mask: 4294967295
 srcIP Mask: 0.0.0.0
 dstIP Mask: 255.255.255.255
 srcPort Mask: 0
 dstPort Mask: 0
 proto Mask: 0
 pdrID: 0
 fseID 805306368
 fseidIP 0
 ctrID: 0
 farID: 2
 qerID: 2
 needDecap: 1
 allocIPFlag: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:76 main.(*bess).sendMsgToUPF()                                              
 FAR ID: 0
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 0
 applyAction: 2
 tunnelType: 1
 tunnelIP4Src: 3323068417
 tunnelIP4Dst: 184615297
 tunnelTEID: 805306368
 tunnelPort: 2152
 sendEndMarker: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:87 main.(*bess).sendMsgToUPF()                                              
 FAR ID: 1
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 1
 applyAction: 2
 tunnelType: 0
 tunnelIP4Src: 0
 tunnelIP4Dst: 0
 tunnelTEID: 0
 tunnelPort: 0
 sendEndMarker: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:87 main.(*bess).sendMsgToUPF()                                              
 FAR ID: 2
 fseID: 805306368
 fseIDIP: 0
 dstIntf: 1
 applyAction: 2
 tunnelType: 1
 tunnelIP4Src: 3323133953
 tunnelIP4Dst: 218169799
 tunnelTEID: 2415919104
 tunnelPort: 2152
 sendEndMarker: false
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:87 main.(*bess).sendMsgToUPF()                                              
 QER ID: 0
 fseID: 805306368
 QFI: 9
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:98 main.(*bess).sendMsgToUPF()                                              
 QER ID: 1
 fseID: 805306368
 QFI: 8
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:98 main.(*bess).sendMsgToUPF()                                              
 QER ID: 2
 fseID: 805306368
 QFI: 7
 fseIDIP: 0.0.0.0
 Uplink Status: 0
 Downling Status: 0
 Uplink MBR: 50000
 Downlink MBR: 50000
 Uplink GBR: 50000
 Downlink GBR: 50000
TRAC[2021-08-06T05:13:21Z]/pfcpiface/bess.go:98 main.(*bess).sendMsgToUPF()
```

File: pfcpiface/parse-far.go (hunk 1 above)
Comment (krsna1729) [reply pfcpiface/parse-far.go:70]: After this LGTM

File: pfcpiface/main.go [hunk 2]
```diff
@@ -122,15 +123,25 @@ func ParseIP(name string, iface string) net.IP {
...
 	var intf fastPath
 	// read and parse json startup file
 	ParseJSON(configPath, &conf)
-	log.Println(conf)
+
+	// Set up logger
+	log.SetReportCaller(true)
+	log.SetFormatter(&log.TextFormatter{
+		FullTimestamp:   true,
+		TimestampFormat: "",
```
Comment (pudelkoM) [inline pfcpiface/main.go:137]: That line can be removed

**Comment by krsna1729** [comment]
@pudelkoM can you check the CI? looks like the cpiface container which runs this binary is crashing

```
  cpiface:
    Container ID:  docker://0410093f57086772dc50cfb18969945affec737594b3220c4ca33ea90564a49d
    Image:         registry.aetherproject.org/omecproject/upf-epc-pfcpiface:master-PR_311-89cf2b0
    Image ID:      docker-pullable://registry.aetherproject.org/omecproject/upf-epc-pfcpiface@sha256:560e176bdbdc8f0ae8579147305bcc952674e2a7fe245c03f375381a7ed950cb
    Port:          <none>
    Host Port:     <none>
    Command:
      pfcpiface
    Args:
      -config
      /tmp/conf/upf.json
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Fri, 06 Aug 2021 05:19:29 +0000
      Finished:     Fri, 06 Aug 2021 05:19:29 +0000
    Ready:          False
```

**Comment by krsna1729** [comment]
retest this please

**Review by krsna1729 — DISMISSED** [review]

File: pfcpiface/main.go [hunk 3]
```diff
@@ -122,15 +123,24 @@ func ParseIP(name string, iface string) net.IP {
...
 	ParseJSON(configPath, &conf)
-	log.Println(conf)
+
+	// Set up logger
+	log.SetReportCaller(true)
+	log.SetFormatter(&log.TextFormatter{
+		FullTimestamp: true,
+	})
+	if level, err := log.ParseLevel(conf.LogLevel); err != nil {
+		log.Fatalln(err)
```
Comment (krsna1729) [inline pfcpiface/main.go:139]: if `conf.LogLevel` is nil, set it to info and then do this. Currently if I do not have `log_level` in my config, its exiting.

```
FATA[2021-08-07T07:36:23Z]/pfcpiface/main.go:139 main.main() not a valid logrus Level: ""
```

File: pfcpiface/parse-pdr.go [hunk 4]
```diff
@@ -47,33 +49,34 @@ func needAllocIP(ueIPaddr *ie.UEIPAddressFields) bool {
...
+	fmt.Println(&b, "tunnelIP4DstMask:", int2ip(p.tunnelIP4DstMask))
+	fmt.Println(&b, "tunnelTEIDMaskMask:", p.tunnelTEIDMask)
+	fmt.Println(&b, "srcIPMask:", int2ip(p.srcIPMask))
+	fmt.Println(&b, "dstIPMask:", int2ip(p.dstIPMask))
+	fmt.Println(&b, "srcPortMask:", p.srcPortMask)
+	fmt.Println(&b, "dstPortMask:", p.dstPortMask)
+	fmt.Println(&b, "protoMask:", p.protoMask)
+	fmt.Println(&b, "pdrID:", p.pdrID)
+	fmt.Println(&b, "fseID", p.fseID)
+	fmt.Println(&b, "fseidIP", int2ip(p.fseidIP))
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:73]: lets fix missing `:` too

File: pfcpiface/parse-pdr.go [hunk 5]
```diff
@@ -47,33 +49,34 @@ func needAllocIP(ueIPaddr *ie.UEIPAddressFields) bool {
...
+	fmt.Println(&b, "tunnelIP4Dst:", int2ip(p.tunnelIP4Dst))
+	fmt.Println(&b, "tunnelTEID:", p.tunnelTEID)
+	fmt.Println(&b, "srcIP:", int2ip(p.srcIP))
+	fmt.Println(&b, "dstIP:", int2ip(p.dstIP))
+	fmt.Println(&b, "srcPort:", p.srcPort)
+	fmt.Println(&b, "dstPort:", p.dstPort)
+	fmt.Println(&b, "proto:", p.proto)
+	fmt.Println(&b, "srcIfaceMask:", p.srcIfaceMask)
+	fmt.Println(&b, "tunnelIP4DstMask:", int2ip(p.tunnelIP4DstMask))
+	fmt.Println(&b, "tunnelTEIDMaskMask:", p.tunnelTEIDMask)
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:65]: typo MaskMask

File: pfcpiface/parse-pdr.go [hunk 6]
```diff
@@ -47,33 +49,34 @@ func needAllocIP(ueIPaddr *ie.UEIPAddressFields) bool {
...
+	var b = strings.Builder{}
+	fmt.Println(&b, "srcIface:", p.srcIface)
+	fmt.Println(&b, "tunnelIP4Dst:", int2ip(p.tunnelIP4Dst))
+	fmt.Println(&b, "tunnelTEID:", p.tunnelTEID)
+	fmt.Println(&b, "srcIP:", int2ip(p.srcIP))
+	fmt.Println(&b, "dstIP:", int2ip(p.dstIP))
+	fmt.Println(&b, "srcPort:", p.srcPort)
+	fmt.Println(&b, "dstPort:", p.dstPort)
+	fmt.Println(&b, "proto:", p.proto)
+	fmt.Println(&b, "srcIfaceMask:", p.srcIfaceMask)
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:63]:
```suggestion
	fmt.Fprintf(&b, "srcIfaceMask:%#x\n", p.srcIfaceMask)
```
Similar for all non ip subnet masks

File: pfcpiface/parse-pdr.go [hunk 7]
```diff
@@ -47,33 +49,34 @@ func needAllocIP(ueIPaddr *ie.UEIPAddressFields) bool {
...
-	log.Println("qerID:", p.qerID)
-	log.Println("needDecap:", p.needDecap)
-	log.Println("allocIPFlag:", p.allocIPFlag)
-	log.Println("--------------------------------------------")
+// Satisfies the fmt.Stringer interface
+func (p pdr) String() string {
+	var b = strings.Builder{}
+	fmt.Println(&b, "srcIface:", p.srcIface)
+	fmt.Println(&b, "tunnelIP4Dst:", int2ip(p.tunnelIP4Dst))
+	fmt.Println(&b, "tunnelTEID:", p.tunnelTEID)
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:57]: hex

File: pfcpiface/parse-pdr.go [hunk 8]
```diff
@@ -47,33 +49,34 @@ func needAllocIP(ueIPaddr *ie.UEIPAddressFields) bool {
...
+	fmt.Println(&b, "srcIfaceMask:", p.srcIfaceMask)
+	fmt.Println(&b, "tunnelIP4DstMask:", int2ip(p.tunnelIP4DstMask))
+	fmt.Println(&b, "tunnelTEIDMaskMask:", p.tunnelTEIDMask)
+	fmt.Println(&b, "srcIPMask:", int2ip(p.srcIPMask))
+	fmt.Println(&b, "dstIPMask:", int2ip(p.dstIPMask))
+	fmt.Println(&b, "srcPortMask:", p.srcPortMask)
+	fmt.Println(&b, "dstPortMask:", p.dstPortMask)
+	fmt.Println(&b, "protoMask:", p.protoMask)
+	fmt.Println(&b, "pdrID:", p.pdrID)
+	fmt.Println(&b, "fseID", p.fseID)
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:72]: hex

File: pfcpiface/parse-far.go [hunk 9]
```diff
@@ -52,20 +54,21 @@ type far struct {
...
+	var b = strings.Builder{}
+	fmt.Println(&b, "farID:", f.farID)
+	fmt.Println(&b, "fseID:", f.fseID)
+	fmt.Println(&b, "fseIDIP:", int2ip(f.fseidIP))
+	fmt.Println(&b, "dstIntf:", f.dstIntf)
+	fmt.Println(&b, "applyAction:", f.applyAction)
+	fmt.Println(&b, "tunnelType:", f.tunnelType)
+	fmt.Println(&b, "tunnelIP4Src:", int2ip(f.tunnelIP4Src))
+	fmt.Println(&b, "tunnelIP4Dst:", int2ip(f.tunnelIP4Dst))
+	fmt.Println(&b, "tunnelTEID:", f.tunnelTEID)
```
Comment (krsna1729) [inline pfcpiface/parse-far.go:68]: hex

File: pfcpiface/parse-far.go [hunk 10]
```diff
@@ -52,20 +54,21 @@ type far struct {
...
-	log.Println("tunnelIP4Dst:", f.tunnelIP4Dst)
-	log.Println("tunnelTEID:", f.tunnelTEID)
-	log.Println("tunnelPort:", f.tunnelPort)
-	log.Println("sendEndMarker:", f.sendEndMarker)
-	log.Println("--------------------------------------------")
+// Satisfies the fmt.Stringer interface
+func (f far) String() string {
+	var b = strings.Builder{}
+	fmt.Println(&b, "farID:", f.farID)
+	fmt.Println(&b, "fseID:", f.fseID)
```
Comment (krsna1729) [inline pfcpiface/parse-far.go:61]: hex

File: pfcpiface/parse-qer.go [hunk 11]
```diff
@@ -22,20 +24,22 @@ type qer struct {
...
-	log.Println("Uplink MBR:", q.ulMbr)
-	log.Println("Downlink MBR:", q.dlMbr)
-	log.Println("Uplink GBR:", q.ulGbr)
-	log.Println("Downlink GBR:", q.dlGbr)
-	log.Println("--------------------------------------------")
+// Satisfies the fmt.Stringer interface
+func (q qer) String() string {
+	var b = strings.Builder{}
+	fmt.Println(&b, "qerID:", q.qerID)
+	fmt.Println(&b, "fseID:", q.fseID)
```
Comment (krsna1729) [inline pfcpiface/parse-qer.go:31]: hex

File: pfcpiface/main.go (hunk 3 above)
Comment (pudelkoM) [reply pfcpiface/main.go:144]: I added that check to `ParseJSON`

Comment (krsna1729) [reply pfcpiface/main.go:144]: CI is passing now. That must have been the cause for its failure 😀

**Comment by pudelkoM** [comment]
I think I got most/all hex prints, and rebased on latest master.

File: pfcpiface/parse-far.go [hunk 12]
```diff
@@ -52,20 +54,21 @@ type far struct {
...
+	var b = strings.Builder{}
+	fmt.Println(&b, "farID:", f.farID)
+	fmt.Fprintf(&b, "fseID: %x\n", f.fseID)
+	fmt.Println(&b, "fseIDIP:", int2ip(f.fseidIP))
+	fmt.Println(&b, "dstIntf:", f.dstIntf)
+	fmt.Println(&b, "applyAction:", f.applyAction)
+	fmt.Println(&b, "tunnelType:", f.tunnelType)
+	fmt.Println(&b, "tunnelIP4Src:", int2ip(f.tunnelIP4Src))
+	fmt.Println(&b, "tunnelIP4Dst:", int2ip(f.tunnelIP4Dst))
+	fmt.Fprintf(&b, "tunnelTEID: %x", f.tunnelTEID)
```
Comment (krsna1729) [inline pfcpiface/parse-far.go:68]: Missing newline

```suggestion
	fmt.Fprintf(&b, "tunnelTEID: %x\n", f.tunnelTEID)
```

**Review by krsna1729 — APPROVED** [review]

**Pull request merged**